DDP Communication hook: (Patch) Fix the way we pass future result to buckets.

sinannasir opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43734 DDP Communication hook: (Patch) Fix the way we pass future result to buckets.**

Following the additional GH comments on the original PR https://github.com/pytorch/pytorch/pull/43307.

Differential Revision: [D23380288](https://our.internmc.facebook.com/intern/diff/D23380288/)

sinannasir committed:
DDP Communication hook: (Patch) Fix the way we pass future result to buckets.

Following the additional GH comments on the original PR https://github.com/pytorch/pytorch/pull/43307.

Differential Revision: [D23380288](https://our.internmc.facebook.com/intern/diff/D23380288/)

[ghstack-poisoned]

rohan-varma reviewed (commented):
Looks good overall, though I'm not sure if this will actually improve perf for the initialize_buckets case, since we still do a copy? Is there any way to avoid that (such as make the bucket_view_out point to bucket_view_in)?

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -774,25 +774,31 @@ void Reducer::initialize_bucket_views(
+    // by default `bucket_views_out` and `bucket_views_in` are
+    // essentially the same thing.
+    if (populate_bucket_views_in) {
+      replica.bucket_views_out = replica.bucket_views_in;
```
This makes the code simpler and easier to digest, although I'm concerned this will still incur an unnecessary copy. Do we need to this copy or can they both point to the same vector in the non-comm hook case?

rohan-varma reviewed (approved):
Per offline discussion, in non-comm hook case this is only called once, and thus avoiding the copy isn't expected to provide too big a win. We could still look into it as a later enhancement, but stamping this for now.  Feel free to rebase to clear up CI

sinannasir committed:
Update on "DDP Communication hook: (Patch) Fix the way we pass future result to buckets."


Following the additional GH comments on the original PR https://github.com/pytorch/pytorch/pull/43307.

Differential Revision: [D23380288](https://our.internmc.facebook.com/intern/diff/D23380288/)

[ghstack-poisoned]

sinannasir committed:
Update on "DDP Communication hook: (Patch) Fix the way we pass future result to buckets."


Following the additional GH comments on the original PR https://github.com/pytorch/pytorch/pull/43307.

Differential Revision: [D23380288](https://our.internmc.facebook.com/intern/diff/D23380288/)

[ghstack-poisoned]
